src/Arcus.WebApi.Correlation/CorrelationMiddleware.cs

```diff
@@ -0,0 +1,133 @@
+                    httpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
+                    await httpContext.Response.WriteAsync($"No correlation transaction ID request header '{_options.Transaction.HeaderName}' was allowed in the request");
+
+                    return;
+                }
+
+                _logger.LogTrace("Correlation request header '{HeaderName}' found with transaction ID '{TransactionId}'", _options.Transaction.HeaderName, transactionIds);
+            }
+
+            string operationId = DetermineOperationId(httpContext);
+            string transactionId = DetermineTransactionId(httpContext, transactionIds);
+            var correlation = new CorrelationInfo(operationId, transactionId);
+            httpContext.Features.Set(correlation);
+
+            AddCorrelationResponseHeaders(httpContext, operationId, transactionId);
+
+            // TODO: on exception, should we enrich the exception.Data with the correlation info?
+            await _next(httpContext);
+        }
+
+        private static string DetermineOperationId(HttpContext httpContext)
+        {
+            // TODO: make ID generation configurable for the consumer.
+            return httpContext.TraceIdentifier ?? Guid.NewGuid().ToString();
+        }
+
+        private string DetermineTransactionId(HttpContext httpContext, StringValues transactionIds)
+        {
+            // TODO: make ID generation configurable for the consumer.
+            if (String.IsNullOrWhiteSpace(transactionIds.ToString()))
+            {
+                if (_options.Transaction.GenerateWhenNotSpecified)
+                {
+                    return Guid.NewGuid().ToString();
+                }
+                else
+                {
+                    return null;
+                }
+            }
+            else
+            {
+                return transactionIds.ToString();
+            }
+        }
+
+        private void AddCorrelationResponseHeaders(HttpContext httpContext, string operationId, string transactionId)
+        {
+            if (_options.Operation.IncludeInResponse)
+            {
+                httpContext.Response.OnStarting(() =>
+                {
+                    TryAddResponseHeader(httpContext, _options.Operation.HeaderName, operationId);
+                    return Task.CompletedTask;
+                });
+            }
+
+            if (_options.Transaction.IncludeInResponse)
+            {
+                httpContext.Response.OnStarting(() =>
+                {
+                    TryAddResponseHeader(httpContext, _options.Transaction.HeaderName, transactionId);
+                    return Task.CompletedTask;
+                });
+            }
+        }
+
+        private void TryAddResponseHeader(HttpContext httpContext, string headerName, string headerValue)
+        {
+            if (!httpContext.Response.Headers.ContainsKey(headerName))
```

Comment:
Should we check of just overwrite? I'd say the latter, no?

Reply:
Yeah, makes sense. I'll change that.